### ISSUE TITLE
Defining new primitive: MPI and their handles

### ISSUE THREAD

**Issue by PhilipVinc** [google/jax#3221]
Dear everyone, I have been trying to support MPI with Jax with success by using Cython and adding a new primitive.

However, my first trial used an hardcoded World comm and reduction operations.
I wanted to make this a bit more general, but I have some problems, so I hope you can answer me here...

I want to support the mpi operation `MPI.Allreduce(MPI.IN_PLACE, arr, operation, comm)`.
`comm` and `operation` are two custom classes, but all I really need to pass down to the C kernel is their handle.

So, first I define a new primitive and the function 
```python
mpi_allreduce_p = core.Primitive("sum_inplace_mpi")  # Create the primitive

def jax_mpi_allreduce(x, op, comm):
    # MPI_Comm and MPI_Op are not primitive XLA types
    # however, they are simple opaque pointers to MPI structs
    # so we can pass their pointer value as a uint64
    op_ptr = _np.uint64(_MPI._handleof(op))
    comm_ptr = _np.uint64(_MPI._handleof(comm))
    return mpi_allreduce_p.bind(x, op_ptr, comm_ptr)
```
Note: I think my error is here, because Id like to encode them, or bind them, as compile time constants...

then I define the abstract eval, which does nothing
```python
def mpi_allreduce_abstract_eval(xs, op, comm):
    return abstract_arrays.ShapedArray(xs.shape, xs.dtype)
```

now finally, the encoding
```python
# Helper functions
def _constant_s32_scalar(c, x):
    return _ops.Constant(c, _np.int32(x))
def _unpack_builder(c):
    # If `c` is a ComputationBuilder object, extracts the underlying XlaBuilder.
    return getattr(c, "_builder", c)

#  This function compiles the operation
def mpi_allreduce_xla_encode(c, x, op, comm):
    c = _unpack_builder(c)
    x_shape = c.GetShape(x)
    dtype = x_shape.element_type()
    dims = x_shape.dimensions()
    # compute total number of elements in array
    nitems = dims[0]
    for el in dims[1:]:
        nitems *= el
    # those kernels have been loaded through cython.
        # those kernels have been loaded through cython.
    if dtype == _np.float32:
        kernel = b"mpi_allreduce_f32"
    elif dtype == _np.float64:
        kernel = b"mpi_allreduce_f64"
    elif dtype == _np.complex64:
        kernel = b"mpi_allreduce_c64"
    elif dtype == _np.complex128:
        kernel = b"mpi_allreduce_c128"

    print(op)
    print(comm)

    return _ops.CustomCall(
        c,
        kernel,
        operands=(xla_client.ops.Constant(c, _np.int32(nitems)), 
                  xla_client.ops.Constant(c, op), 
                  xla_client.ops.Constant(c, comm), 
                  x),
        shape=xla_client.Shape.array_shape(dtype, dims),
    )
```
If I run the above code I get the error 
```python
  File "/Users/filippovicentini/Dropbox/Dottorato/2-Algoritmi-Originali/mpi4xla/mpi_xla.py", line 106, in mpi_allreduce_xla_encode
    xla_client.ops.Constant(c, op),
RuntimeError: Invalid argument: Unknown NumPy type O size 8
```

In the code above `op` and `comm` will be of type `jaxlib.xla_extension.XlaOp`, but I would expect them to be compile time constant uint64 values.

How do I access those and encode them in the CustomCall?

**Comment by dionhaefner** [comment]
:wave: @PhilipVinc, would you mind sharing your code? I had a similar idea and would like to see if I can make some progress.

**Comment by PhilipVinc** [comment]
https://gist.github.com/PhilipVinc/3574a8adf88a4cff13a7c154afac9f1e

**Comment by PhilipVinc** [comment]
I would love to make a package, I had started working on it, but unless google folks help me here on how to pass pointers around I am at a dead end.

The code above I linked only works on MPI_WORLD_COMM

**Comment by dionhaefner** [comment]
That looks really nice, thanks!

I took the liberty to adapt this to work on my machine with current JAX versions and dump it here:

https://github.com/dionhaefner/jax-mpi

If you were planning to make a package, I can transfer ownership of the repository to you.

In any case, I'll see if I can make any progress on the COMM issue when I'm back from vacation in 2 weeks. But even having something that works with COMM_WORLD could be very valuable.

**Comment by PhilipVinc** [comment]
Don't worry about the package. I wanted to make a package about this, but in order to support arbitrary types/operations/communicators you need to be able to pass their bare pointers to xla. 
The pointers are not hard to obtain, but I don't understand how to serialise them (see this code if you want to play where I left off https://gist.github.com/PhilipVinc/c98c8e14dfac55c8ab948fe95b8a9618 ).

I got stuck there. And as I have basically no time those days, I stopped working on it.
If you manage to get this to work it would be nice.

BTW, the code I posted only works for cpus. to get it to work to gpus a little bit of work is needed (not much though).

**Comment by dionhaefner** [comment]
I think I've got it. Apparently, when calling `bind` on the primitive, only *positional arguments* are passed as JAX tracers / `JaxOP`, not *keyword arguments*. So this works:

```python
def sum_inplace_jax_primitive(x, comm):
    comm_ptr = _np.uint64(MPI._handleof(comm))
    return sum_inplace_p.bind(x, comm=comm_ptr)  # instead of bind(x, comm_ptr)
```

[Working prototype](https://github.com/dionhaefner/jax-mpi/blob/master/test.py)

**Comment by PhilipVinc** [comment]
AMAZING! It would be lovely if this could be documented @froystig

With that I can let go of my ugly workaround!

I have uploaded [MpI4jax](https://github.com/PhilipVinc/mpi4jax) to Github. 
I took the liberty to list you as a contributor.

The only operation it supports for now is Allreduce, but it can work with arbitrary ops and communicators and mixed with batches and AD, which is something that I need.

I'd like to register it soon in PyPy/Conda because I'll be adding it as a dependency of another project I'm working on.
If it's okay for you we can focus there contributions?
